plugins/ltac/extratactics.ml4

```diff
@@ -950,6 +950,22 @@ TACTIC EXTEND is_const
     | _ -> Tacticals.New.tclFAIL 0 (Pp.str "not a constant") ]
 END;;
 
+let rec is_literal sigma t =
+  match EConstr.kind sigma t with
+  | Ind _ -> true
+  | Construct _ -> true
+  | App(f, args) -> is_literal sigma f && Array.for_all (is_literal sigma) args
+  | _ -> false
```

Comment:
It is a bit weird that `0 :: nil` is a literal and that `nat` is a literal but `nat :: nil` is not (because `Type` is not and `::` has an implicit)

Reply:
So sorts should be considered as litterals, right?

Reply:
I don't have a strong opinion on if nat should be a literal or not. My main use of is_literal is to avoid rewriting a literal with something else. E.g. if I have an equation which after some destructs ends up to be 1=x to avoid that 1 is replaced with x in some automated rewrite with hypothesis. Obviously it usually wouldn't make a lot of sense to rewrite nat = T either.

Still it also would make sense to keep a connection to what people usually call a literal in other languages. Many languages have a notion of list or other structures literals, but I I don't think that if you ask some people (without this context) if they think a type name is a literal, they would probably say no.

So in summary: I am open to treat nat is literals in the tactic we discuss here, but I woudln't call it is_literal then, but maybe is_groundterm.

Maybe it makes more sense to have is_groundterm and if someone want's is_literal in the strict sense, he can still use is_groundterm and check if the head symbol is a constructor.

In this case I would probably use is_gorundterm directly in my tests.